File: leetcode/1_two_sum.cpp
// https://leetcode.com/problems/two-sum

class Solution {
public:
    vector<int> twoSum(vector<int>& nums, int target) {
        vector<int> result;
        
        if (nums.size() < 2) return result;

        vector<int> indices(nums.size());
        int iota = 0;
        for (auto& e : indices) e = iota++;

        sort(indices.begin(), indices.end(), [&nums](int i, int j){ return nums[i] < nums[j]; });
        
        int i = 0;
        int j = nums.size() - 1;

        for (int i = 0, j = nums.size() - 1; i < j; )
        {
            int candidate = nums[indices[i]] + nums[indices[j]];
            if (candidate == target) return {indices[i], indices[j]};
            if (candidate > target) --j; else ++i;
        }
    }
};

File: leetcode/1_two_sum.cs
public class Solution {
    public int[] TwoSum(int[] nums, int target) {
        int[] indices = Enumerable.Range(0, nums.Length).OrderBy(e => nums[e]).ToArray();
        int i = 0, j = nums.Length - 1;
        while (j > i)
        {
            int sum = nums[indices[i]] + nums[indices[j]];
            if (sum > target) --j;
            else if (sum < target) ++i;
            else break;
        }
        return new[]{ indices[i], indices[j] };
    }
}
